File: ios/OpenEOSControl/App/LiveViewMonitoring.swift
import CoreGraphics
import Foundation
import ImageIO
import UIKit

enum LiveViewFrameGuide: String, CaseIterable, Identifiable, Sendable {
    case off
    case ratio16x9
    case ratio2x39
    case ratio1x1
    case ratio4x3

    var id: String { rawValue }

    var aspectRatio: CGFloat? {
        switch self {
        case .off: nil
        case .ratio16x9: 16.0 / 9.0
        case .ratio2x39: 2.39
        case .ratio1x1: 1
        case .ratio4x3: 4.0 / 3.0
        }
    }
}

enum LiveViewDesqueeze: String, CaseIterable, Identifiable, Sendable {
    case off
    case x1_33
    case x1_5
    case x1_8
    case x2

    var id: String { rawValue }

    var horizontalScale: CGFloat {
        switch self {
        case .off: 1
        case .x1_33: 1.33
        case .x1_5: 1.5
        case .x1_8: 1.8
        case .x2: 2
        }
    }
}

struct LiveViewMonitorSettings: Equatable, Hashable, Sendable {
    var histogramVisible = false
    var zebraThresholdPercent: Int?
    var falseColorEnabled = false
    var focusPeakingEnabled = false
    var frameGuide = LiveViewFrameGuide.off
    var safeAreaVisible = false
    var desqueeze = LiveViewDesqueeze.off

    var needsPixelAnalysis: Bool {
        histogramVisible || zebraThresholdPercent != nil || falseColorEnabled || focusPeakingEnabled
    }
}

struct LiveViewMonitorAnalysis: Equatable, Sendable {
    let width: Int
    let height: Int
    let histogram: [Int]
    let overlayRGBA: [UInt8]?

    func overlayImage() -> UIImage? {
        guard let overlayRGBA else { return nil }
        let data = Data(overlayRGBA) as CFData
        guard let provider = CGDataProvider(data: data),
              let image = CGImage(
                width: width,
                height: height,
                bitsPerComponent: 8,
                bitsPerPixel: 32,
                bytesPerRow: width * 4,
                space: CGColorSpaceCreateDeviceRGB(),
                bitmapInfo: CGBitmapInfo(
                    rawValue: CGBitmapInfo.byteOrder32Big.rawValue | CGImageAlphaInfo.last.rawValue
                ),
                provider: provider,
                decode: nil,
                shouldInterpolate: false,
                intent: .defaultIntent
              ) else {
            return nil
        }
        return UIImage(cgImage: image)
    }
}

func liveViewAnalysisDimensions(width: Int, height: Int) -> (width: Int, height: Int) {
    precondition(width > 0 && height > 0)
    let scale = min(
        1,
        min(Double(maximumAnalysisWidth) / Double(width), Double(maximumAnalysisHeight) / Double(height))
    )
    return (
        max(1, Int((Double(width) * scale).rounded())),
        max(1, Int((Double(height) * scale).rounded()))
    )
}

func analyzeLiveViewData(_ data: Data, settings: LiveViewMonitorSettings) -> LiveViewMonitorAnalysis? {
    guard settings.needsPixelAnalysis,
          let imageSource = CGImageSourceCreateWithData(data as CFData, nil),
          let source = CGImageSourceCreateThumbnailAtIndex(
            imageSource,
            0,
            [
                kCGImageSourceCreateThumbnailFromImageAlways: true,
                kCGImageSourceCreateThumbnailWithTransform: true,
                kCGImageSourceThumbnailMaxPixelSize: maximumAnalysisDecodeSize,
            ] as CFDictionary
          ) else {
        return nil
    }
    let dimensions = liveViewAnalysisDimensions(width: source.width, height: source.height)
    var rgba = [UInt8](repeating: 0, count: dimensions.width * dimensions.height * 4)
    let rendered = rgba.withUnsafeMutableBytes { bytes -> Bool in
        guard let context = CGContext(
            data: bytes.baseAddress,
            width: dimensions.width,
            height: dimensions.height,
            bitsPerComponent: 8,
            bytesPerRow: dimensions.width * 4,
            space: CGColorSpaceCreateDeviceRGB(),
            bitmapInfo: CGBitmapInfo.byteOrder32Big.rawValue | CGImageAlphaInfo.premultipliedLast.rawValue
        ) else {
            return false
        }
        context.interpolationQuality = .medium
        context.translateBy(x: 0, y: CGFloat(dimensions.height))
        context.scaleBy(x: 1, y: -1)
        context.draw(
            source,
            in: CGRect(x: 0, y: 0, width: dimensions.width, height: dimensions.height)
        )
        return true
    }
    guard rendered else { return nil }
    return analyzeLiveViewPixels(
        rgba: rgba,
        width: dimensions.width,
        height: dimensions.height,
        zebraThresholdPercent: settings.zebraThresholdPercent,
        focusPeakingEnabled: settings.focusPeakingEnabled,
        falseColorEnabled: settings.falseColorEnabled
    )
}

func analyzeLiveViewPixels(
    rgba: [UInt8],
    width: Int,
    height: Int,
    zebraThresholdPercent: Int?,
    focusPeakingEnabled: Bool,
    falseColorEnabled: Bool = false
) -> LiveViewMonitorAnalysis {
    precondition(width > 0 && height > 0)
    precondition(rgba.count == width * height * 4)
    precondition(zebraThresholdPercent == nil || (50...100).contains(zebraThresholdPercent!))

    let pixelCount = width * height
    var luminance = [Int](repeating: 0, count: pixelCount)
    var histogram = [Int](repeating: 0, count: histogramBucketCount)
    for index in 0..<pixelCount {
        let offset = index * 4
        let value = (
            54 * Int(rgba[offset])
                + 183 * Int(rgba[offset + 1])
                + 19 * Int(rgba[offset + 2])
        ) >> 8
        luminance[index] = value
        histogram[min(histogramBucketCount - 1, value * histogramBucketCount / 256)] += 1
    }

    let needsOverlay = zebraThresholdPercent != nil || falseColorEnabled || focusPeakingEnabled
    var overlay: [UInt8]?
    let zebraLuminance = zebraThresholdPercent.map { $0 * 255 / 100 }

    if needsOverlay {
        var overlayPixels = [UInt8](repeating: 0, count: pixelCount * 4)
        for y in 0..<height {
            for x in 0..<width {
                let index = y * width + x
                if falseColorEnabled {
                    writeOverlayColor(falseColor(luminance[index]), at: index, into: &overlayPixels)
                }
                if let zebraLuminance, luminance[index] >= zebraLuminance {
                    let color = ((x + y) / 2) % 4 < 2 ? zebraLight : zebraDark
                    writeOverlayColor(color, at: index, into: &overlayPixels)
                }
            }
        }
        if focusPeakingEnabled, width >= 3, height >= 3 {
            for y in 1..<(height - 1) {
                for x in 1..<(width - 1) {
                    let index = y * width + x
                    let gradient = abs(luminance[index + 1] - luminance[index - 1])
                        + abs(luminance[index + width] - luminance[index - width])
                    if gradient >= focusPeakingGradient {
                        writeOverlayColor(focusPeakingColor, at: index, into: &overlayPixels)
                    }
                }
            }
        }
        overlay = overlayPixels
    }

    return LiveViewMonitorAnalysis(
        width: width,
        height: height,
        histogram: histogram,
        overlayRGBA: overlay
    )
}

private func writeOverlayColor(_ color: (UInt8, UInt8, UInt8, UInt8), at index: Int, into rgba: inout [UInt8]) {
    let offset = index * 4
    rgba[offset] = color.0
    rgba[offset + 1] = color.1
    rgba[offset + 2] = color.2
    rgba[offset + 3] = color.3
}

private func falseColor(_ luminance: Int) -> (UInt8, UInt8, UInt8, UInt8) {
    switch luminance {
    case ..<16: (80, 22, 122, 224)
    case ..<40: (32, 61, 160, 224)
    case ..<75: (41, 182, 209, 224)
    case ..<140: (76, 175, 80, 224)
    case ..<190: (242, 209, 61, 224)
    case ..<235: (242, 140, 40, 224)
    default: (232, 59, 53, 224)
    }
}

private let maximumAnalysisWidth = 120
private let maximumAnalysisHeight = 80
private let maximumAnalysisDecodeSize = max(maximumAnalysisWidth, maximumAnalysisHeight)
private let histogramBucketCount = 64
private let focusPeakingGradient = 72
private let zebraLight: (UInt8, UInt8, UInt8, UInt8) = (255, 255, 255, 176)
private let zebraDark: (UInt8, UInt8, UInt8, UInt8) = (0, 0, 0, 120)
private let focusPeakingColor: (UInt8, UInt8, UInt8, UInt8) = (40, 197, 217, 255)

File: android/app/src/main/java/dev/openeos/control/ui/LiveViewMonitoring.kt
package dev.openeos.control.ui

import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.Rect
import kotlin.math.abs
import kotlin.math.min
import kotlin.math.roundToInt

enum class LiveViewFrameGuide(val aspectRatio: Float?) {
    OFF(null),
    RATIO_16_9(16f / 9f),
    RATIO_2_39(2.39f),
    RATIO_1_1(1f),
    RATIO_4_3(4f / 3f),
}

enum class LiveViewDesqueeze(val horizontalScale: Float) {
    OFF(1f),
    X1_33(1.33f),
    X1_5(1.5f),
    X1_8(1.8f),
    X2(2f),
}

data class LiveViewMonitorSettings(
    val histogramVisible: Boolean = false,
    val zebraThresholdPercent: Int? = null,
    val falseColorEnabled: Boolean = false,
    val focusPeakingEnabled: Boolean = false,
    val frameGuide: LiveViewFrameGuide = LiveViewFrameGuide.OFF,
    val safeAreaVisible: Boolean = false,
    val desqueeze: LiveViewDesqueeze = LiveViewDesqueeze.OFF,
) {
    val needsPixelAnalysis: Boolean
        get() = histogramVisible || zebraThresholdPercent != null || falseColorEnabled || focusPeakingEnabled
}

internal data class LiveViewMonitorAnalysis(
    val width: Int,
    val height: Int,
    val histogram: IntArray,
    val overlayPixels: IntArray?,
)

internal fun analyzeLiveViewBitmap(
    bitmap: Bitmap,
    zebraThresholdPercent: Int?,
    focusPeakingEnabled: Boolean,
    falseColorEnabled: Boolean = false,
): LiveViewMonitorAnalysis {
    val dimensions = analysisDimensions(bitmap.width, bitmap.height)
    val sampled = Bitmap.createBitmap(dimensions.first, dimensions.second, Bitmap.Config.ARGB_8888)
    Canvas(sampled).drawBitmap(
        bitmap,
        null,
        Rect(0, 0, dimensions.first, dimensions.second),
        Paint(Paint.ANTI_ALIAS_FLAG or Paint.FILTER_BITMAP_FLAG),
    )
    return try {
        val pixels = IntArray(dimensions.first * dimensions.second)
        sampled.getPixels(pixels, 0, dimensions.first, 0, 0, dimensions.first, dimensions.second)
        analyzeLiveViewPixels(
            pixels = pixels,
            width = dimensions.first,
            height = dimensions.second,
            zebraThresholdPercent = zebraThresholdPercent,
            focusPeakingEnabled = focusPeakingEnabled,
            falseColorEnabled = falseColorEnabled,
        )
    } finally {
        sampled.recycle()
    }
}

internal fun analyzeLiveViewPixels(
    pixels: IntArray,
    width: Int,
    height: Int,
    zebraThresholdPercent: Int?,
    focusPeakingEnabled: Boolean,
    falseColorEnabled: Boolean = false,
): LiveViewMonitorAnalysis {
    require(width > 0 && height > 0)
    require(pixels.size == width * height)
    require(zebraThresholdPercent == null || zebraThresholdPercent in 50..100)

    val luminance = IntArray(pixels.size)
    val histogram = IntArray(HISTOGRAM_BUCKETS)
    pixels.forEachIndexed { index, pixel ->
        val red = pixel shr 16 and 0xff
        val green = pixel shr 8 and 0xff
        val blue = pixel and 0xff
        val value = (54 * red + 183 * green + 19 * blue) shr 8
        luminance[index] = value
        histogram[min(HISTOGRAM_BUCKETS - 1, value * HISTOGRAM_BUCKETS / 256)]++
    }

    val overlay = if (zebraThresholdPercent != null || falseColorEnabled || focusPeakingEnabled) {
        IntArray(pixels.size)
    } else {
        null
    }
    val zebraLuminance = zebraThresholdPercent?.let { it * 255 / 100 }

    if (overlay != null) {
        for (y in 0 until height) {
            for (x in 0 until width) {
                val index = y * width + x
                if (falseColorEnabled) overlay[index] = falseColor(luminance[index])
                if (zebraLuminance != null && luminance[index] >= zebraLuminance) {
                    overlay[index] = if (((x + y) / 2) % 4 < 2) ZEBRA_LIGHT else ZEBRA_DARK
                }
            }
        }
        if (focusPeakingEnabled && width >= 3 && height >= 3) {
            for (y in 1 until height - 1) {
                for (x in 1 until width - 1) {
                    val index = y * width + x
                    val gradient = abs(luminance[index + 1] - luminance[index - 1]) +
                        abs(luminance[index + width] - luminance[index - width])
                    if (gradient >= FOCUS_PEAKING_GRADIENT) overlay[index] = FOCUS_PEAKING_COLOR
                }
            }
        }
    }

    return LiveViewMonitorAnalysis(width, height, histogram, overlay)
}

private fun falseColor(luminance: Int): Int = when {
    luminance < 16 -> 0xE050167A.toInt()
    luminance < 40 -> 0xE0203DA0.toInt()
    luminance < 75 -> 0xE029B6D1.toInt()
    luminance < 140 -> 0xE04CAF50.toInt()
    luminance < 190 -> 0xE0F2D13D.toInt()
    luminance < 235 -> 0xE0F28C28.toInt()
    else -> 0xE0E83B35.toInt()
}

private fun analysisDimensions(width: Int, height: Int): Pair<Int, Int> {
    require(width > 0 && height > 0)
    val scale = min(1f, min(MAX_ANALYSIS_WIDTH.toFloat() / width, MAX_ANALYSIS_HEIGHT.toFloat() / height))
    return (width * scale).roundToInt().coerceAtLeast(1) to
        (height * scale).roundToInt().coerceAtLeast(1)
}

private const val MAX_ANALYSIS_WIDTH = 120
private const val MAX_ANALYSIS_HEIGHT = 80
private const val HISTOGRAM_BUCKETS = 64
private const val FOCUS_PEAKING_GRADIENT = 72
private const val ZEBRA_LIGHT = 0xB0FFFFFF.toInt()
private const val ZEBRA_DARK = 0x78000000
private const val FOCUS_PEAKING_COLOR = 0xFF28C5D9.toInt()
